### DISCUSSION TITLE
Why am i getting this "signin?error=Callback" error?

### DISCUSSION THREAD

**Discussion by ChrisB007** [nextauthjs/next-auth#2781] [Help] [answered]
Hello friends,
I am trying to configure my app to using Prisma + next-auth, and I followed the instructions from both websites to a 'T'. However, the only error I am getting is this one "Try signing with a different account." displayed on the browser right after you click on the GitHub sign-in button.  Below are my codes:

- The error states

```
[next-auth][error][oauth_callback_handler_error] 
https://next-auth.js.org/errors#oauth_callback_handler_error TypeError: adapter.getAdapter is not a function
 ```

- From my API Next-auth file

``` 
import NextAuth from "next-auth";
import Providers from "next-auth/providers";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

export default NextAuth({
  adapter: PrismaAdapter(prisma),
  providers: [
    Providers.GitHub({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,

      scope: "read:user",
    }),
  ],

  database: {
    type: "mongodb",
    uri: process.env.DATABASE_URL,
    w: "majority",
    useNewUrlParser: true,
    useUnifiedTopology: true,
    retryWrites: true,
    synchronize: true,
  },

  secret: process.env.SECRET,

  session: {
    jwt: true,
  },
  jwt: {
    signingKey: process.env.JWT_SIGNING_PRIVATE_KEY,
  },


  debug: false,
});
```

-  My schema.prisma file

``` 
datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb"]
}

```

- From my .env file

```  

DATABASE_URL="mongodb+srv://userMe:yVTCusZsdmyMM3MJ@publictrades.ic2ix.mongodb.net/MyDBades?retryWrites=true&w=majority"

 ```

- From my console

```  

Request URL: http://localhost:3000/api/auth/signin?error=Callback
Request Method: GET
Status Code: 304 Not Modified
Remote Address: [::1]:3000
Referrer Policy: strict-origin-when-cross-origin

```

**Comment by balazsorban44** [comment, marked as answer]
You are using a new adapter with v3 of next-auth.

have a look at the migration docs https://next-auth.js.org/getting-started/upgrade-v4

how did you install the adapter and next-auth?

**Comment by ryanbarr** [comment]
@balazsorban44 I also have this issue. I followed the instructions for installing `next-auth` and the Prisma adapter that are listed on the [Prisma Adapter Page](https://next-auth.js.org/adapters/prisma).

Unfortunately, following the `npm install` instructions as provided will error out immediately due to an unresolvable dependency. It seems as though installing `@next-auth/prisma-adapter@next` is looking to resolve a peer at `next-auth@">=4"`. Beta versions do not meet this criteria as they are not considered a major release, to my knowledge. As a result, `4.0.0-beta.1` and `4.0.0-beta.2` will not resolve.

In my situation, having this been the first time I've worked with `next-auth`, what would you recommend? Should I manually downgrade to v3 and use the legacy adapter?

**Reply by balazsorban44** [reply]
It depends. beta is pretty stable IMO and we don't expect major breaking changes, but we are currently waiting on a security audit and some adapters/providers need migrating to v4 (see open PRs on the adapters repo or #2524) 

If you just started a new project, I would go with v4. not sure about the install errors though, I haven't encountered them myself. 😬

**Reply by ryanbarr** [reply]
Appreciate the quick response! The error should be consistently reproducible in any fresh Next.js app, from what I'm seeing.

Using: `npm install next-auth@beta @prisma/client @next-auth/prisma-adapter@next`

Result:
```
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: **CENSORED**
npm ERR! Found: next-auth@4.0.0-beta.2
npm ERR! node_modules/next-auth
npm ERR!   next-auth@"4.0.0-beta.2" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer next-auth@">=4" from @next-auth/prisma-adapter@0.5.2-next.15
npm ERR! node_modules/@next-auth/prisma-adapter
npm ERR!   @next-auth/prisma-adapter@"0.5.2-next.15" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

It has no problem finding `4.0.0-beta.2`, but to my knowledge, due to semantic versioning, `4.0.0-beta.2` comes _before_ the `4.0.0` major release and therefore does not satisfy the `>=4` requirement set. My understanding is that [requiring `next-auth@latest`](https://github.com/nextauthjs/adapters/blob/%40next-auth/prisma-adapter%400.5.2/package.json#L39) is still seeing `4.0.0` as the latest, even though it's deprecated.

I'd love to use v4 and the Prisma adapter. I'm using the official Twitch provider that, according to #2524 has been tested and confirmed working. I would love to help find the discrepancy here because following the documentation provided on the official site does not result in a functional implementation currently, at least from my end.

**Reply by balazsorban44** [reply]
The discrepancy was me publishing 4.0.0 by accident, which I cannot remove from npm anymore... 🙄 https://next-auth.js.org/getting-started/upgrade-v4#verify-the-correct-version

I have created many v4 setups in the past few months, and never had the problem you are describing, might be something special with you package manager. I've tested with NPM 6, 7, and yarn 1.

Could you please add a reproduction anyway, even if it's super simple, and list all packages/libraries/tools that you think might be relevant (node, npm, yarn, etc)

cc @ndom91 any ideas?

**Reply by ryanbarr** [reply]
No worries, @balazsorban44 -- I understand what happened and empathize, I've done the same! If you plan on keeping `4.0.0` deprecated, my suggestion would be bump to 4.1.0, or 4.0.1 depending on how you define a minor release. Technically your `beta.1` and `beta.2` builds are version for a `4.0.0` release which is inaccurate because you can no longer re-publish `4.0.0`.

To clarify: **v4 itself is not the issue**. This issue is very specific to the `@next` tagged release and latest release of `@next-auth/prisma-adapter`. It has a peer dependency of `>=4` which, for whatever reason, it cannot resolve. My guess currently is due to its deprecation. Normally you'd get a warning saying the package is deprecated and to upgrade to a newer version, but there is no newer version available because the `beta.1` and `beta.2` tags are _technically_ "older" than `4.0.0` because of how semantic versioning treats those tags.

My current setup is pretty straightforward, all of my dependencies are:

```
    "@next-auth/prisma-adapter": "^0.5.4",
    "@prisma/client": "^3.0.2",
    "@types/next-auth": "^3.15.0",
    "@types/react": "^17.0.20",
    "next": "^11.1.2",
    "next-auth": "^4.0.0-beta.2",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "swr": "^1.0.1",
    "typescript": "^4.4.3"
```

I was able to work around the issue by doing the following, in order:

```bash
npm install next-auth@beta
npm install @prisma/client
npm install @next-auth/prisma-adapter@0.5.4 --force
npm install @types/next-auth
prisma generate
```

This just forces `@next-auth/prisma-adapter` to ignore dependency version requirements, since it believes it needs `>=4` and thinks that is not being met by `^4.0.0-beta.2`.

**Reply by ndom91** [reply]
Yeah so npm7 is more strict about this and auto installs peer deps, I'm pretty sure you could skip around the prisma '>=4' issue and also use the original "all-in-one" install cmd and pass the --force or --legacy-peer-deps flag
